operator/cmd/supervisor/supervisor.go

```diff
@@ -0,0 +1,159 @@
+package supervisor
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"os/signal"
+	"time"
+
+	"github.com/redpanda-data/redpanda-operator/pkg/otelutil/log"
+	"github.com/spf13/cobra"
+	"golang.org/x/time/rate"
+)
+
+func Command() *cobra.Command {
+	var retryBurst int
+	var retryRate time.Duration
+	var gracePeriod time.Duration
+
+	cmd := &cobra.Command{
+		Use:     "supervisor command...",
+		Example: "supervisor --burst-rate=1m -- redpanda-operator sidecar --redpanda-yaml path/on/disk",
+		Short: `supervisor is a stand in for Kubernetes' Sidecar feature.
+It continuously re-runs the provided command, forwarding signals, in a loop until SIGINT is received.
+Subprocess retries are rate limited with token bucket that may be tuned via the --retry-burst and --retry-rate flags.
+The default retry parameters are tuned for a a long running process, such as the sidecar subcommand.`,
+		Args: cobra.MinimumNArgs(1),
+		Run: func(cmd *cobra.Command, args []string) {
+			ctx, cancel := signal.NotifyContext(cmd.Context(), os.Interrupt)
+			defer cancel()
+
+			signals := make(chan os.Signal, 16)
+
+			signal.Notify(signals)
+
+			limiter := rate.NewLimiter(rate.Every(retryRate), retryBurst)
+			runner := Runner{
+				Command:     args,
+				GracePeriod: gracePeriod,
+				Signals:     signals,
+				Stdout:      os.Stdout,
+				Stderr:      os.Stderr,
+			}
+
+			// Loop forever, re-running our child process until the parent
+			// process receives a request to shutdown.
+			for {
+				// We start by consuming a retry "credit" for correctness but
+				// also to fail quickly if the provided retry parameters result
+				// in an unusable limiter.
+				reservation := limiter.Reserve()
+
+				// The API and docs are a bit confusing here as they mention
+				// parameters that's not exposed. OK indicates whether or not
+				// the reservation COULD be granted not if it was. i.e. OK will
+				// be false if retryBurst is < 1.
+				if !reservation.OK() {
+					panic(fmt.Sprintf("invalid retry parameters:\n--retry-burst=%d\n--retry-rate=%s", retryBurst, retryRate))
+				}
+
+				delay := reservation.Delay()
+
+				if delay > 0 {
+					log.Info(ctx, "backing off retries", "delay", delay)
+				}
+
+				// Otherwise wait until we have a retry "credit" available.
+				select {
+				case <-ctx.Done():
```

Comment:
```suggestion
				case <-ctx.Done():
					return
```
otherwise you'll spawn a child process unnecessarily.

Reply:
Fixed 😅